### ISSUE TITLE
Numba "randomly" deduces wrong types

### ISSUE THREAD

**Issue by pwuertz** [numba/numba#2365]
I'm getting invalid results from a JIT compiled function after upgrading numba. I tried to reduce the code as much as possible, but even small and meaningless changes to code will "un-trigger" the bug.

The last line before return swaps the upper and lower bits of an uint8. When using numba, this line produces an incorrect result, but only in this exact situation. Seems like a JIT compiler bug:
```python
import numpy as np
import numba as nb

def test_swap():
    Q = np.ones(1, dtype=np.uint8)
    for i in range(1):
        p = nb.uint8(1)
        p |= bool(Q[i])
        p = nb.uint8((p << 4) | (p >> 4))
        return p

assert test_swap() == nb.jit(test_swap)()
```
For example, changing the array lookup in line 8 from `Q[i]` to `Q[0]` shouldn't affect the outcome of this function in any way, yet it "fixes" the problem. Omitting the loop, which does basically nothing, fixes it. Wrapping the bool in line 8 with nb.uint8, which shouldn't change anything, fixes it.

EDIT: In those cases where the function fails Numba deduces `float64` for variable `p` instead of integer types. This of course fails when bit shifting `p`. So the question is why Numba sometimes deduces `float64`.

**Comment by pwuertz** [comment]
Commit f94c12023e8d8994376f06f5d1948dce7cf06bd9 is the first version where this function is failing. The change does affect bit shifts, but I don't see why it would explain this erratic behavior. Maybe there is another JIT bug that has something to do with variable/register lifetime tracking?

**Comment by stuartarchibald** [comment]
Thanks for the report. I can replicate on `0.29`, but the latest release `0.33` seems fine, could you use the latest release perhaps?

**Comment by pwuertz** [comment]
I did try with the (by then) current master. Version string is `0.34.0.dev0+291.gd33dbb475-py2.7-linux-x86_64.egg`. My guess is that the bisected commit I mentioned only triggered the bug for my specific code, other commits may hide and re-trigger it again. The fundamental problem is probably hidden somewhere else? The randomness of this behavior really confuses me. I had a look at `inspect_types`, and it seems to be messed up now and then.

**Comment by seibert** [comment]
Right after we released version 0.33.0, we merged some fairly large compiler changes to master that may have introduced a bug.  We'll see if we can track this down, and for now, downgrading to the current release (0.33.0).

**Comment by pwuertz** [comment]
Just checked out `0.33.0` and getting the same error.

**Comment by stuartarchibald** [comment]
Thanks for the extra information. This appears to effect Python 2.7 only.

**Comment by sklam** [comment]
I can replicate this on py3.5.

This looks like an issue with casting and the sequence of casting ends up with `float64 | float64` and a bug that we accepted bitwise on real numbers.

**Comment by stuartarchibald** [comment]
Looking at this further.
```python
def test_or(a, b):
    return a | b

z = nb.njit(test_or)
res = z(np.float64(1), np.float64(2))
z.inspect_types()
print("result: %s" % res)
```
gives:
```python
--------------------------------------------------------------------------------
# File: issue2365.py
# --- LINE 22 --- 
# label 0
#   del b
#   del a
#   del $0.3

def test_or(a, b):

    # --- LINE 23 --- 
    #   a = arg(0, name=a)  :: float64
    #   b = arg(1, name=b)  :: float64
    #   $0.3 = a | b  :: bool
    #   $0.4 = cast(value=$0.3)  :: bool
    #   return $0.4

    return a | b


================================================================================
result: True
```
which is due to the types accepted by the bitwise OR intrinsic (integer domain types) being reachable in the typemap from a `float64`, using complex FP types correctly raises a `TypeError`.

**Comment by pwuertz** [comment]
I changed the topic because (after #2382) the code I posted throws an exception instead of returning invalid results.

However, there is still a bug in Numba causing "random" float64 type deductions.

**Comment by stuartarchibald** [comment]
Thanks for reporting back. The intention of https://github.com/numba/numba/pull/2382 was to raise if an invalid cast was happening, so it's good news it is now doing that. Could you please describe the bug you are now seeing, along with a reproducer if you have it? Perhaps open a new ticket if it isn't related specifically to the current issue? Thanks.

**Comment by pwuertz** [comment]
As @sklam described earlier there are actually two bugs my initial example suffered from. The primary bug is a wrong type deduction (float64) which exposed a second bug where Numba accepted float64 as bit shift operand. The second bug got fixed, but the primary bug remains where Numba does a spurious cast to `float64` for apparently no reason.

The code to reproduce the error is the same, but be aware that Numba now deduces pyobjects instead of c types if `nopython=True` is not set. Here is the new reproducer:
```python
import numpy as np
import numba as nb

@nb.jit(nopython=True)
def test_swap():
    Q = np.ones(1, dtype=np.uint8)
    for i in range(1):
        p = nb.uint8(1)
        p |= bool(Q[i])
        p = nb.uint8((p << 4) | (p >> 4))
        return p

test_swap()
```
Numba wrongly deduces `float64` for variable `p`, which then gives you the new exception at the bit shift operations. What confuses me the most is that this behavior appears to be completely unpredictable. Apparently meaningless changes to the code will make Numba deduce `int64` for `p` (which works, although I'd expect Numba to deduce `uint8` for `p`). For example:
- Change `p |= bool(Q[i])` to `p |= bool(Q[0])`
- Remove one of the two bit shift operations

**Comment by stuartarchibald** [comment]
I've added a patch here https://github.com/numba/numba/pull/2424 which means that the type inference will be:
 * Consistent across Python versions.
 * Performed using a constraint system that is assembled and operated on in the order of the labels in the IR.

This should reduce the impact of seemingly "meaningless" changes and use outer block information to have greater influence on inner.

I hope to look at potentially altering the type inference mechanism to use narrower types, but have some concerns relating to performance.

**Comment by pwuertz** [comment]
Fixed by #2424
